### PR TITLE
[CIS-1145] Unit Tests For Filter Decoding

### PR DESCRIPTION
**Pull request by kojiba**
[-] Replaced deprecated throws of NSError
    with correct DecodingErrors
[-] Improved Warnings in JSONDecoder and Encoder Tests
[-] Moved reusable test logic to FilterTestExtensions_Tests
[-] Renamed TestScope to FilterTestScope
[-] Updated Changelog according to changes
[-] Updated Package.swift with new tests ignored

Test plan: 
- run all unit tests succeed
- quickly run through demo app, see everything work as usual

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -27,6 +27,7 @@ A new `ChatChannelVC` is introduced that represents the old `ChatMessageListVC`,
 
 ### 🔄 Changed
 - Nuke dependency was updated to v10 [#1405](https://github.com/GetStream/stream-chat-swift/pull/1405)
+- Instead of deprecated `NSError()` on a filter decoding now correct DecodingError thrown [#1417](https://github.com/GetStream/stream-chat-swift/pull/1417/)
```
Comment (tbarbugli) [inline CHANGELOG.md:30]: we dont have to update changelog for something like this

File: Package.swift [hunk 2]
```diff
@@ -191,7 +191,9 @@ var streamChatSourcesExcluded: [String] { [
     "Workers/ChannelUpdater_Mock.swift",
     "Workers/EventNotificationCenter_Tests.swift",
     "Workers/ChatClientUpdater_Tests.swift",
+    "Query/FilterTestExtensions_Tests.swift",
```
Comment (tbarbugli) [inline Package.swift:194]: just to be sure, you did not edit this manually, right?

Comment (kojiba) [reply Package.swift:194]: Sure, runned script from Scripts

File: Sources/StreamChat/Query/FilterDecoding_Tests.swift [hunk 3]
```diff
@@ -0,0 +1,221 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+@testable import StreamChat
+import XCTest
+
+class FilterDecoding_Tests: XCTestCase {
```
Comment (tbarbugli) [inline Sources/StreamChat/Query/FilterDecoding_Tests.swift:8]: please make sure to test that encoding and decoding are symmetrical so that we dont end up with encoded data failing at decoding

it must not be possible to encode data that cannot be decoded

File: Sources/StreamChat/Query/FilterDecoding_Tests.swift [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    func testFilterRightSideDecodingThrowExceptionWithoutOperationKey() {
+        let filterJson = #"{ "test_key": { "eq":"test_value_1" } }"#
+
+        checkDecodingFilterThrowException(json: filterJson)
+    }
+
+    func testFilterRightSideDecodingThrowExceptionOnNonExistingOperation() {
+        let filterJson = #"{ "test_key": { "$myspecialOperation":"test_value_1" } }"#
```
Comment (tbarbugli) [inline Sources/StreamChat/Query/FilterDecoding_Tests.swift:35]: Let's be 100% sure that it's impossible for the filter to end up encoded this way (ie. if we have something like .customOperator("myspecialOperation") we are in trouble)

**Review by tbarbugli — COMMENTED** [review]
see my comments

File: Package.swift (hunk 2 above)
Comment (b-onc) [reply Package.swift:194]: You shouldn't run this script manually. Please run `./bootstrap.sh`, it'll install this script as a commit hook, so it runs automatically.

File: Sources/StreamChat/Query/FilterDecoding_Tests.swift (hunk 4 above)
Comment (kojiba) [reply Sources/StreamChat/Query/FilterDecoding_Tests.swift:35]: Yes, for sure, but for decoder, we should test this corner case to be thrown exception and not crash

File: Sources/StreamChat/Query/Filter.swift [hunk 5]
```diff
@@ -368,10 +372,22 @@ private struct FilterRightSide: Decodable {
...
+            throw DecodingError.dataCorruptedError(
+                forKey: container.allKeys.last ?? ArbitraryKey(""),
+                in: container,
+                debugDescription: "FilterRightSide keys count should be only 1"
+            )
+        }
         
         let key = container.allKeys.first!
-        guard key.stringValue.hasPrefix("$") else { throw NSError() }
+        guard key.stringValue.hasPrefix("$") else {
```
Comment (evsaev) [inline Sources/StreamChat/Query/Filter.swift:384]:
```suggestion
        guard let operator = container.allKeys.first?.stringValue, operator.hasPrefix("$") else { ... }
        self.operator = operator
```

**Review by evsaev — APPROVED** [review]
👏 🚢

**Review by nuno-vieira — APPROVED** [review]
LGTM ✅
